### PR TITLE
filtering on value

### PR DESCRIPTION
**Pull request by pavlopoul**

### CODE REVIEW

**Comment by prestoncarman** [comment]
Is this ready for review? Did you get the filter on value working?
Also, please include a rule comment at the top outlining the before and after state of the plan.

**Comment by pavlopoul** [comment]
I need to remove some code but other than that it is ready...I fixed the problem with search query

**Comment by prestoncarman** [comment]
Go ahead and remove the code and add the rule comments, then I will review the code.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 1]
```diff
@@ -199,6 +200,8 @@
         normalization.add(new RemoveUnusedAssignAndAggregateRule());
         normalization.add(new PushValueIntoDatascanRule());
         normalization.add(new PushKeysOrMembersIntoDatascanRule());
+        normalization.add(new EliminateSubplanForSingleItemsRule());
+        //normalization.add(new PushIndexingValueIntoDatascanRule());
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java]: Should this rule be uncommented? I don't see PushIndexingValueIntoDatascanRule being added to the ruleset.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 2]
```diff
@@ -350,6 +354,7 @@
         // Re-infer all types, so that, e.g., the effect of not-is-null is
         // propagated.
         prepareForJobGenRewrites.add(new PushProjectIntoDataSourceScanRule());
+        //prepareForJobGenRewrites.add(new PushIndexingValueIntoDatascanRule());
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java]: Remove commented out lines.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/EliminateSubplanForSingleItemsRule.java [hunk 3]
```diff
@@ -141,43 +142,52 @@ public boolean rewritePre(Mutable<ILogicalOperator> opRef, IOptimizationContext
...
         // Ensure input is from a UNNEST operator.
         AbstractLogicalOperator subplanInput = (AbstractLogicalOperator) subplan.getInputs().get(0).getValue();
-        if (subplanInput.getOperatorTag() != LogicalOperatorTag.ASSIGN) {
-            return false;
-        }
-        AssignOperator assign = (AssignOperator) subplanInput;
-        if (!assign.getVariables().contains(vre2.getVariableReference())) {
-            return false;
-        }
+        if (!(subplanInput.getOperatorTag() == LogicalOperatorTag.ASSIGN
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/EliminateSubplanForSingleItemsRule.java]: Is the fix here to allow for the DATASCAN operator?

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushIndexingValueIntoDatascanRule.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        }
+        AbstractLogicalOperator op1 = (AbstractLogicalOperator) opRef.getValue();
+        if (op1.getOperatorTag() != LogicalOperatorTag.SELECT) {
+            return false;
+        }
+        SelectOperator select = (SelectOperator) op1;
+
+        op2 = (AbstractLogicalOperator) select.getInputs().get(0).getValue();
+
+        if (op2.getOperatorTag() != LogicalOperatorTag.EXCHANGE) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushIndexingValueIntoDatascanRule.java]: Can this rule run before the physical rules are run?

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java [hunk 5]
```diff
@@ -68,6 +69,7 @@ public VXQueryIndexingOperatorDescriptor(IOperatorDescriptorRegistry spec, VXQue
         totalDataSources = (short) ds.getTotalDataSources();
         recordDescriptors[0] = rDesc;
         childSeq = ds.getChildSeq();
+        indexSeq=ds.getIndexSeq();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/VXQueryIndexingOperatorDescriptor.java]: formatting.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 6]
```diff
@@ -134,7 +159,14 @@ public void init() throws SystemException {
 
         int partition = ctx.getTaskAttemptId().getTaskId().getPartition();
         ITreeNodeIdProvider nodeIdProvider = new TreeNodeIdProvider((short) partition);
-        handler = new SAXContentHandler(false, nodeIdProvider, appender, childSequenceTypes);
+        List<SequenceType> newChildSequenceType=new ArrayList<SequenceType>();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: formatting.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 7]
```diff
@@ -201,24 +250,29 @@ private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) thr
...
-            handler.endElement(uri, firstBit, firstBit);
-
+            while (dots <firstColon) {
+                nextdot = dots;
+                dots = contents.indexOf(".", dots + 1);
+                element = contents.substring(nextdot + 1, dots);
+                if (dots > firstColon) {
+                    element = contents.substring(nextdot + 1, firstColon);
+                }
+                List<String> names = new ArrayList<>();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: We should be able to run this code by creating a single names array and just clear with each pass. I think the correct way to fix this is to add the arrays to the IndexAttributes object with a reset function. Then the whole class could have one Attributes object that is reset before each use. Less object creation and more object reuse.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 8]
```diff
@@ -251,6 +305,11 @@ private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) thr
             handler.endElement(uri, lastBit, lastBit);
 
         }
+        
+        while(elements>0){
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: formatting.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 9]
```diff
@@ -201,24 +250,29 @@ private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) thr
         int lastDot = contents.lastIndexOf('.');
         String type = contents.substring(lastDot + 1);
         String lastBit = contents.substring(firstColon + 1, lastDot);
-
+        int dots = contents.indexOf(".");
+        int nextdot;
+        String element="";
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: formatting.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 10]
```diff
@@ -74,27 +77,49 @@
...
-            childLocalName[index] = FunctionHelper.getStringFromBytes(nameTest.getLocalName());
 
+            NodeType nodeType = (NodeType) sType.getItemType();
+            if (nodeType.getNodeKind() == NodeKind.ELEMENT) {
+                ElementType eType = (ElementType) nodeType;
+                NameTest nameTest = eType.getNameTest();
+                childLocalName[index] = FunctionHelper.getStringFromBytes(nameTest.getLocalName());
+            } else {
+                if (!indexSeq.isEmpty()) {
+                    byte[] indexBytes = new byte[indexSeq.get(0).length];
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: Is there a byte copy function that could be used here?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 11]
```diff
@@ -74,27 +77,49 @@
...
+        this.indexSeq = indexSeq;
         this.indexName = indexPath;
         this.appender = appender;
         final DynamicContext dCtx = (DynamicContext) ctx.getJobletContext().getGlobalJobData();
         childSequenceTypes = new ArrayList<>();
         for (int typeCode : childSeq) {
             childSequenceTypes.add(dCtx.getStaticContext().lookupSequenceType(typeCode));
         }
+        TextType tt = TextType.INSTANCE;
+        if (!indexSeq.isEmpty()) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: Why add something to childSequenceType? Can you not have a separate process for index that happens after childSequenceType processing?

**Review by prestoncarman — COMMENTED** [review]
I think we need to discuss some of these changes in more detail. Also, I am not sure the rule is activated.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/EliminateSubplanForSingleItemsRule.java [hunk 12]
```diff
@@ -140,44 +143,54 @@ public boolean rewritePre(Mutable<ILogicalOperator> opRef, IOptimizationContext
...
+
+            IntegerPointable pTypeCode = (IntegerPointable) IntegerPointable.FACTORY.createPointable();
+            tvp.getValue(pTypeCode);
+            SequenceType sType = dCtx.lookupSequenceType(pTypeCode.getInteger());
+            if (sType.getQuantifier() != Quantifier.QUANT_ONE) {
+                return false;
+            }
+        }
+
+        //        if (!assign.getVariables().contains(vre2.getVariableReference())) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/EliminateSubplanForSingleItemsRule.java]: Remove commented out code.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 13]
```diff
@@ -190,6 +192,7 @@
         normalization.add(new RemoveRedundantDataExpressionsRule());
         normalization.add(new RemoveRedundantPromoteExpressionsRule());
         normalization.add(new RemoveRedundantCastExpressionsRule());
+
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java]: Remove new line.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java [hunk 14]
```diff
@@ -227,6 +231,7 @@
     public static final List<IAlgebraicRewriteRule> buildTypeInferenceRuleCollection() {
         List<IAlgebraicRewriteRule> typeInfer = new LinkedList<>();
         typeInfer.add(new InferTypesRule());
+
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/RewriteRuleset.java]: Remove new line.

File: vxquery-core/src/main/java/org/apache/vxquery/metadata/AbstractVXQueryDataSource.java [hunk 15]
```diff
@@ -98,6 +100,22 @@ public void addChildSeq(int integer) {
...
+    
+    public void addIndexChildSeq(int integer) {
+        indexChildSeq.add(integer);
+    }
+
+    public List<Integer> getIndexChildSeq() {
+        return indexChildSeq;
+    }
+    
+    public void addIndexSeq(Byte[] index) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/metadata/AbstractVXQueryDataSource.java]: Should this be called addIndexValueSeq? To clarify this is for index values, not paths.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 16]
```diff
@@ -191,8 +262,13 @@ public void parse(ArrayBackedValueStorage abvsFileNode) throws IOException {
     }
 
     private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) throws SAXException {
+        List<String> names = new ArrayList<>();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: Move these arrays into IndexAtributes. Then create a single IndexAttributes object that you call reset before using.

File: vxquery-core/src/main/java/org/apache/vxquery/index/IndexAttributes.java [hunk 17]
```diff
@@ -32,6 +32,12 @@
 
     public IndexAttributes(List<String> n, List<String> v, List<String> u, List<String> l, List<String> t,
             List<String> q) {
+        n.clear();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/index/IndexAttributes.java]: Almost ... move these clear statements to a reset method. Initial the arrays in the constructor. Then the object will no longer have any arguments to the constructor.

**Review by prestoncarman — CHANGES_REQUESTED** [review]
After these changes I think its ready to merge. Steven do you have any additional comments?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java [hunk 18]
```diff
@@ -227,15 +295,8 @@ private int buildElement(ArrayBackedValueStorage abvsFileNode, int fieldNum) thr
...
-            List<String> names = new ArrayList<>();
-            List<String> values = new ArrayList<>();
-            List<String> uris = new ArrayList<>();
-            List<String> localNames = new ArrayList<>();
-            List<String> types = new ArrayList<>();
-            List<String> qNames = new ArrayList<>();
-            whereIFinish = findAttributeChildren(whereIFinish, names, values, uris, localNames, types, qNames);
-            Attributes atts = new IndexAttributes(names, values, uris, localNames, types, qNames);
-
+            whereIFinish = findAttributeChildren(whereIFinish, atts);
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/VXQueryIndexReader.java]: I think reset should be before findAttributeChildren.

**Review by prestoncarman — COMMENTED** [review]
Do we have tests for this new functionality? Which queries in XTest will test this new functionality?

**Comment by pavlopoul** [comment]
All the test cases that have a prefix useindex test the new functionality....however, the only way to verify that the rule actually works is by debugging. I checked that and it works fine.

**Comment by prestoncarman** [comment]
I see you added tests for attributes and indexing. Does this mean that the indexing works for attributes without any changes?

**Comment by pavlopoul** [comment]
Yes, it works....the only change I made was that instead of storing the length of the attribute in the constructor of IndexAttributes (was always 0 since the constructor always initializes the lists) I store it one of the methods where I add elements in the lists.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushIndexingIntoDatascanRule.java [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+            context.addToDontApplySet(this, opRef.getValue());
+            return true;
+        }
+
+        return false;
+
+    }
+
+    private boolean updateDataSource(IVXQueryDataSource dataSource, Mutable<ILogicalExpression> expression) {
+        if (!dataSource.usingIndex()) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/PushIndexingIntoDatascanRule.java:113]: Split into three functions (children, value, attributes).

**Pull request closed without merging**